File: iteration-two-task2/frontend/src/api/modeEval.ts
import client from "./client";

export interface RealtimeSession {
    session_id: string;
    status: string;
    agent_name: string;
    task: string;
    started_at: string;
    ended_at?: string;
    step_count: number;
    tool_call_count: number;
    consecutive_same_action: number;
    node_switch_count: number;
    event_count: number;
    current_node?: string;
    current_agent_type?: string;
    latest_decision: string;
    alert_count: number;
}

export interface RealtimeAlert {
    id: string;
    session_id: string;
    rule: string;
    severity: string;
    message: string;
    decision: string;
    created_at: string;
}

export interface RealtimeStreamToolCall {
    name: string;
    params?: Record<string, unknown>;
    dangerous?: boolean;
}

export interface RealtimeStreamEvent {
    id: string;
    session_id: string;
    event_type: string;
    node: string;
    agent_type: string;
    step: number;
    thought: string;
    message: string;
    next_agent: string;
    tool_calls: RealtimeStreamToolCall[];
    state: Record<string, unknown>;
    duration_sec: number;
    decision: string;
    reason: string;
    source: string;
    created_at: string;
}

export interface RealtimeEventIngestResponse {
    session: RealtimeSession;
    event: RealtimeStreamEvent;
    decision: string;
    reason: string;
    should_stop: boolean;
}

export interface OfflineEvalReport {
    session_id: string;
    task: string;
    success: boolean;
    scores: Record<string, number>;
    metrics: Record<string, number>;
    root_cause: string;
    recommendations: string[];
}

export interface OfflineEvalJob {
    job_id: string;
    status: string;
    created_at: string;
    updated_at: string;
    error?: string;
    report?: OfflineEvalReport;
}

export interface BatchEvalSummary {
    dataset_id: string;
    total_traces: number;
    success_rate: number;
    average_scores: Record<string, number>;
    root_cause_distribution: Record<string, number>;
    sample_reports: OfflineEvalReport[];
}

export interface BatchEvalJob {
    job_id: string;
    status: string;
    created_at: string;
    updated_at: string;
    error?: string;
    summary?: BatchEvalSummary;
    progress?: { processed: number; total: number };
    eval_task_id?: number;
}

export const startRealtimeSession = async (payload: {
    session_id?: string;
    agent_name: string;
    task: string;
    thresholds?: Record<string, number>;
}) => {
    const { data } = await client.post<RealtimeSession>("/mode-realtime/sessions/start", payload);
    return data;
};

export const realtimeStepStart = async (sessionId: string, payload: { step?: number; thought?: string; action?: string }) => {
    const { data } = await client.post<RealtimeSession>(`/mode-realtime/sessions/${sessionId}/step-start`, payload);
    return data;
};

export const realtimeToolCall = async (sessionId: string, payload: { tool_name: string; params?: Record<string, unknown>; dangerous?: boolean }) => {
    const { data } = await client.post<RealtimeSession>(`/mode-realtime/sessions/${sessionId}/tool-call`, payload);
    return data;
};

export const realtimeStepEnd = async (sessionId: string, payload: { step?: number; action?: string; duration_sec?: number; step_result?: string }) => {
    const { data } = await client.post<RealtimeSession>(`/mode-realtime/sessions/${sessionId}/step-end`, payload);
    return data;
};

export const endRealtimeSession = async (sessionId: string) => {
    const { data } = await client.post<RealtimeSession>(`/mode-realtime/sessions/${sessionId}/end`);
    return data;
};

export const listRealtimeSessions = async () => {
    const { data } = await client.get<{ total: number; items: RealtimeSession[] }>("/mode-realtime/sessions");
    return data;
};

export const getRealtimeSessionDetail = async (sessionId: string) => {
    const { data } = await client.get<{ session: RealtimeSession; latest_alerts: RealtimeAlert[]; latest_events: RealtimeStreamEvent[] }>(`/mode-realtime/sessions/${sessionId}`);
    return data;
};

export const ingestRealtimeStreamEvent = async (
    sessionId: string,
    payload: {
        event_type?: string;
        node: string;
        agent_type?: string;
        step?: number;
        thought?: string;
        message?: string;
        next_agent?: string;
        tool_calls?: RealtimeStreamToolCall[];
        state?: Record<string, unknown>;
        duration_sec?: number;
        timestamp?: string;
        source?: string;
    }
) => {
    const { data } = await client.post<RealtimeEventIngestResponse>(`/mode-realtime/sessions/${sessionId}/events`, payload);
    return data;
};

export const submitOfflineEval = async (trace: Record<string, unknown>) => {
    const { data } = await client.post<OfflineEvalJob>("/mode-offline/jobs/submit", { trace });
    return data;
};

export const listOfflineJobs = async () => {
    const { data } = await client.get<{ total: number; items: OfflineEvalJob[] }>("/mode-offline/jobs");
    return data;
};

export const getOfflineJob = async (jobId: string) => {
    const { data } = await client.get<OfflineEvalJob>(`/mode-offline/jobs/${jobId}`);
    return data;
};

export const runBatchEval = async (datasetId: string, strategyName?: string) => {
    const payload: Record<string, unknown> = { dataset_id: datasetId };
    if (strategyName) payload.strategy_name = strategyName;
    const { data } = await client.post<BatchEvalJob>("/mode-batch/jobs/run", payload);
    return data;
};

export const listBatchJobs = async () => {
    const { data } = await client.get<{ total: number; items: BatchEvalJob[] }>("/mode-batch/jobs");
    return data;
};

export const getBatchJob = async (jobId: string) => {
    const { data } = await client.get<BatchEvalJob>(`/mode-batch/jobs/${jobId}`);
    return data;
};

File: iteration-two-task2/backend/app/schemas/mode_eval.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class RealtimeSessionStartRequest(BaseModel):
    session_id: str | None = None
    agent_name: str = "default-agent"
    task: str = ""
    thresholds: dict[str, float] = Field(default_factory=dict)


class RealtimeStepStartRequest(BaseModel):
    step: int | None = None
    thought: str | None = None
    action: str | None = None


class RealtimeToolCallRequest(BaseModel):
    tool_name: str
    params: dict[str, Any] = Field(default_factory=dict)
    dangerous: bool = False


class RealtimeStepEndRequest(BaseModel):
    step: int | None = None
    action: str | None = None
    duration_sec: float = 0.0
    step_result: str | None = None


class RealtimeStreamToolCall(BaseModel):
    name: str
    params: dict[str, Any] = Field(default_factory=dict)
    dangerous: bool = False


class RealtimeStreamEventRequest(BaseModel):
    event_type: str = "node_execution"
    node: str
    agent_type: str = "worker"
    step: int | None = None
    thought: str | None = None
    message: str | None = None
    next_agent: str | None = None
    tool_calls: list[RealtimeStreamToolCall] = Field(default_factory=list)
    state: dict[str, Any] = Field(default_factory=dict)
    duration_sec: float = 0.0
    timestamp: datetime | None = None
    source: str = "langgraph-stream"


class RealtimeStreamEventRead(BaseModel):
    id: str
    session_id: str
    event_type: str
    node: str
    agent_type: str
    step: int
    thought: str = ""
    message: str = ""
    next_agent: str = ""
    tool_calls: list[dict[str, Any]] = Field(default_factory=list)
    state: dict[str, Any] = Field(default_factory=dict)
    duration_sec: float = 0.0
    decision: str = "CONTINUE"
    reason: str = ""
    source: str = "langgraph-stream"
    created_at: datetime


class RealtimeAlertRead(BaseModel):
    id: str
    session_id: str
    rule: str
    severity: str
    message: str
    decision: str
    created_at: datetime


class RealtimeSessionRead(BaseModel):
    session_id: str
    status: str
    agent_name: str
    task: str
    started_at: datetime
    ended_at: datetime | None = None
    step_count: int
    tool_call_count: int
    consecutive_same_action: int
    node_switch_count: int = 0
    event_count: int = 0
    current_node: str | None = None
    current_agent_type: str | None = None
    latest_decision: str
    alert_count: int


class RealtimeSessionDetailResponse(BaseModel):
    session: RealtimeSessionRead
    latest_alerts: list[RealtimeAlertRead] = Field(default_factory=list)
    latest_events: list[RealtimeStreamEventRead] = Field(default_factory=list)


class RealtimeEventIngestResponse(BaseModel):
    session: RealtimeSessionRead
    event: RealtimeStreamEventRead
    decision: str
    reason: str = ""
    should_stop: bool = False


class RealtimeSessionListResponse(BaseModel):
    total: int
    items: list[RealtimeSessionRead]


class RealtimeAlertsResponse(BaseModel):
    session_id: str
    total: int
    items: list[RealtimeAlertRead]


class StandardTraceStep(BaseModel):
    step: int
    thought: str = ""
    action: dict[str, Any] = Field(default_factory=dict)
    result: str = ""
    duration: float = 0.0
    memory_hit: bool | None = None


class StandardTrace(BaseModel):
    session_id: str
    task: str
    steps: list[StandardTraceStep] = Field(default_factory=list)
    final_result: str = ""
    success: bool | None = None
    expected_result: str | None = None


class OfflineEvalSubmitRequest(BaseModel):
    trace: StandardTrace


class OfflineEvalReport(BaseModel):
    session_id: str
    task: str
    success: bool
    scores: dict[str, float] = Field(default_factory=dict)
    metrics: dict[str, float | int] = Field(default_factory=dict)
    root_cause: str
    recommendations: list[str] = Field(default_factory=list)


class OfflineEvalJobRead(BaseModel):
    job_id: str
    status: str
    created_at: datetime
    updated_at: datetime
    error: str | None = None
    report: OfflineEvalReport | None = None


class OfflineEvalJobListResponse(BaseModel):
    total: int
    items: list[OfflineEvalJobRead]


class DatasetBatchRunRequest(BaseModel):
    dataset_id: str
    strategy_name: str | None = None
    eval_task_id: int | None = None


class DatasetBatchSummary(BaseModel):
    dataset_id: str
    total_traces: int
    success_rate: float
    average_scores: dict[str, float] = Field(default_factory=dict)
    root_cause_distribution: dict[str, int] = Field(default_factory=dict)
    sample_reports: list[OfflineEvalReport] = Field(default_factory=list)


class DatasetBatchJobRead(BaseModel):
    job_id: str
    status: str
    created_at: datetime
    updated_at: datetime
    error: str | None = None
    summary: DatasetBatchSummary | None = None
    progress: dict | None = None
    eval_task_id: int | None = None


class DatasetBatchJobListResponse(BaseModel):
    total: int
    items: list[DatasetBatchJobRead]
